zerver/lib/actions.py

```diff
@@ -5900,3 +5900,61 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
     export.extra_data = ujson.dumps(export_data)
     export.save(update_fields=['extra_data'])
     notify_realm_export(user_profile)
+
+def get_topic_messages(
+    user_profile: UserProfile,
+    stream: Stream,
+    topic_name: str
+) -> List[Message]:
+    log_statsd_event('get_topic_messages')
+    messages = UserMessage.objects.filter(
+        user_profile=user_profile
+    )
+
+    recipient = stream.recipient
+    messages = messages.filter(message__recipient=recipient)
+
+    messages = filter_by_topic_name_via_message(
+        query=messages,
+        topic_name=topic_name,
+    )
+
+    return messages
+
+def do_move_topic_messages(
+    user_profile: UserProfile,
+    stream_origin: Stream,
+    stream_destiny: Stream,
+    topic_name: str
+) -> None:
+    # if they are not public streams don't move any messages
+    if not (stream_origin.is_public() and stream_destiny.is_public()):
+        return
+
+    user_messages = get_topic_messages(user_profile, stream_origin, topic_name)
+    recipient_destiny = get_stream_recipient(stream_destiny.id)
+
+    # We could use a bot for the breadcrump message too
+    internal_send_stream_message(
+        stream_destiny.realm,
+        user_profile,
+        stream_origin,
+        topic_name,
+        "This topic was moved from %s stream" % (
+            stream_origin.name
+        )
+    )
+
+    for user_message in user_messages:
+        user_message.message.recipient = recipient_destiny
+        user_message.message.save(update_fields=['recipient'])
```

Comment:
This isn't going to work for caching and real-time sync reasons.  

See https://zulip.readthedocs.io/en/latest/subsystems/caching.html and https://zulip.readthedocs.io/en/latest/subsystems/events-system.html for details.

Reply:
After some investigation, the caching situation might actually be fine, because `flush_message` ignores `update_fields` currently and just deletes any modified messages from the cache.

Real-time sync is going to be much more problematic; we'll need to extend the interface for "edit message" events to allow the stream to change.  It shouldn't be too bad because the problem is very similar to that of topic editing which we already do that correctly for.  I think ultimately what we'll want to do have this go through the edit message code path.

Reply:
Ok, I'll continue reviewing `do_update_message`. It does a lot of things, I'm not sure if I can reuse that method (adding a stream parameter), because we could have performance issues. I'm going to continue reviewing the code, if I have any questions I'll let you know, thanks